spring-kafka/src/main/java/org/springframework/kafka/mock/MockProducerFactory.java

```diff
@@ -81,14 +99,96 @@ public Producer<K, V> createProducer() {
 
 	@Override
 	public Producer<K, V> createProducer(@Nullable String txIdPrefix) {
-		return txIdPrefix == null && this.defaultTxId == null
-				? this.producerProvider.apply(false, null)
-				: this.producerProvider.apply(true, txIdPrefix == null ? this.defaultTxId : txIdPrefix);
+		return txIdPrefix == null && this.defaultTxId == null ? new CloseSafeMockProducer<>(this.producerProvider.apply(false, null)) :
+				new CloseSafeMockProducer<>(this.producerProvider.apply(true, txIdPrefix == null ? this.defaultTxId : txIdPrefix));
 	}
 
 	@Override
 	public Producer<K, V> createNonTransactionalProducer() {
 		return this.producerProvider.apply(false, null);
 	}
 
+	/**
+	 * A wrapper class for the delegate, inspired by {@link DefaultKafkaProducerFactory.CloseSafeProducer}.
+	 *
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 *
+	 * @author Pawel Szymczyk
+	 */
+	static class CloseSafeMockProducer<K, V> implements Producer<K, V> {
+
+		private static final LogAccessor LOGGER = new LogAccessor(LogFactory.getLog(CloseSafeMockProducer.class));
+
+		private final MockProducer<K, V> delegate;
+
+		CloseSafeMockProducer(MockProducer<K, V> delegate) {
+			this.delegate = delegate;
+		}
+
+		@Override
+		public void initTransactions() {
+			this.delegate.initTransactions();
+		}
+
+		@Override
+		public void beginTransaction() throws ProducerFencedException {
+			this.delegate.beginTransaction();
+		}
+
+		@Override
+		public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets, String consumerGroupId) throws ProducerFencedException {
+			this.delegate.sendOffsetsToTransaction(offsets, consumerGroupId);
+		}
+
+		@Override
+		public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets, ConsumerGroupMetadata groupMetadata) throws ProducerFencedException {
+			this.delegate.sendOffsetsToTransaction(offsets, groupMetadata);
+		}
+
+		@Override
+		public void commitTransaction() throws ProducerFencedException {
+			this.delegate.commitTransaction();
+		}
+
+		@Override
+		public void abortTransaction() throws ProducerFencedException {
+			this.delegate.abortTransaction();
+		}
+
+		@Override
+		public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+			return this.delegate.send(record);
+		}
+
+		@Override
+		public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+			return this.delegate.send(record, callback);
+		}
+
+		@Override
+		public void flush() {
+			this.delegate.flush();
+		}
+
+		@Override
+		public List<PartitionInfo> partitionsFor(String topic) {
+			return this.delegate.partitionsFor(topic);
+		}
+
+		@Override
+		public Map<MetricName, ? extends Metric> metrics() {
+			return this.delegate.metrics();
+		}
+
+		@Override
+		public void close() {
+			close(null);
+		}
+
+		@Override
+		public void close(Duration timeout) {
+			LOGGER.debug(() -> "The closing of delegate producer " + this.delegate + "has been skipped.");
```

Comment:
@pszymczyk While it works for your use case, I wonder if there are other consequences by not closing the underlying producer. Can we only activate this feature (i.e., not closing the producer) based on a property or something along those lines?

Reply:
If you like to provide some configuration I would consider to add something in `KafkaTemplate` [send finally block](https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java#L864). 

Currently, the `KafkaTemplate` provides clients the possibility to pass a custom implementation of the `ProducerFactory`. However, if you pass something other than the `DefaultKafkaProducerFactory` and your implementation does not implement something similar to the `CloseSafeProducer` wrapper, then your `KafkaProducer` will be closed just after the first message is sent, which basically makes no sense.

Reply:
Yes. I find that as a reasonable request.
We already have that `DefaultKafkaProducerFactory` API:
```
	/**
	 * Close the {@link Producer}(s) and clear the cache of transactional
	 * {@link Producer}(s).
	 * @since 2.2
	 */
	@Override
	public void reset() {
```
which lets you control such a lifecycle from the outside of the `KafkaTemplate`.
So, it is going to be safe to not close that cached producer in the `KafkaTemplate` after each send.

This way I suggest to back off the change to the `MockProducerFactory` and raise a new issue which we will consider for the next version.

Thanks

P.S. Removing that my other comment since it contradicts with what we have agreed here.

Reply:
@artembilan sounds really good! I agree that `CloseSafeMockProducer` it's more like a workaround for the real problem I see in `KafkaTemplate` and `DefaultKafkaProducerFactory` indirect relationship, it's great that we all agree on that, I'm closing this PR and I will submit an issue with proper description.